Documentation/articles/domain_driven_design/aggregate_root.md

```diff
@@ -0,0 +1,64 @@
+purposes of a business transaction. The *Aggregate* will maintain all invariants for all objects that comprise the *Aggregate*,
+enforcing all business rules internal to the *Aggregate*.
+
+# Aggregate Root 
+
+All *Aggregates* have a single point of entry that is know as the root.  The *Aggregate Root* is the interface to the external world.
+All interaction with an *Aggregate* is via the *Aggregate Root*.  As such, an *Aggregate Root* **MUST** have a globally unique identifier
+within the system.  Other Entites that are present in the *Aggregate* but are not *Aggregate Roots* require only a locally unique
+identifier, that is, an Id that is unique within the *Aggregate*.  
+
+For example: an *Order* will contain *Order Lines*.  The *Order* is the *Aggregate Root* and must have a unique *Order Id*.  The *Order Lines* cannot be accessed directly.  Any updates to *Order Lines* must be via the *Order*.  That is, you ask the *Order* to make the change to **ITS** *Order Line*.  An *Order Line* is only required to be identifiable with the *Order* therefore it is permissible to have *Order Line Ids* such as 1, 2, 3 etc. on each order.  It is, of course, valid to use a globally unique identifier on each *Order Line* for persistence purposes, while having a locally unique candidate key for presentation, but it is not required.
+
+It is **not permissible** to have references to entities or value objects from an *Aggregate Root* other than the *Aggregate Root* itself.  In a *CQRS* system, it is permissible, indeed required, to have read models that have data from one or more Aggregate Roots.
+
+## Modelling
+
+The modelling of an *Aggregate Root* is based on the invariants that must be maintained when updating.  In our example above, an *Order* will perform aggregate functions on the order lines (e.g. summing the totals, assigning ids, maintaining ordering, etc.) and therefore will potentially update all order lines at the same time.  An *Aggregate* is essentially a business transaction and should include everything that has to be kept consistent.
+
+## System Invariants
+
+The *Aggregate* maintains only its internal invariants. It is not responsible for maintaining invariants between *Aggregates*.
+
+For example, it is not the responsiblity of the individual *Aggregate* to ensure that a *Username* is unique.  It is the responsibility of the **System as a whole** to maintain these system invariants.  The exact mechanism it uses will depend on the invariant and the consequences of breaking them.
+
+{{% notice tip %}}  
+> What you must learn is that these rules are no different that the rules of a computer system. Some of them can be bent. Others can be broken.
+> <cite>Morpheus, The Matrix</cite>
+
+All rules are not created equal.  Business Rules are essentially business problems, not primarily technical problems.  It may be simpler and cheaper
```

Comment:
Udi Dahan has a great example of this.  He's talking about an ATM machine and how it is not absoutely 100% transactionally consistent at all times on the account balance.  So in certain circumstances it's actually possible to withdraw more than your balance.  As devs we look at that as a problem to be solved and can dream up ridiculously complex and expensive solutions, while the business see it as an opportunity... unauthorized overdrafts are expensive and a good source of profit for next to nothing for the bank!